### DISCUSSION TITLE
Pandoc is not installed in ki6 docker image

### DISCUSSION THREAD

**Discussion by cdwilson** [INTI-CMNB/KiBot#228] [KiBot check]
Hi,

I just ran `kibot-check` in the `setsoft/kicad_auto:ki6` docker image and it reported that Pandoc is not installed (output below).

Just wanted to check if it was intentional that pandoc is not included in the `ki6` image, or if this was a bug?

(Otherwise everything checks out OK 👍)

```
# kibot-check
KiBot installation checker

Core:
Linux: 5.10.104 (Linux 2eda77cf3cc9 5.10.104-linuxkit #1 SMP PREEMPT Thu Mar 17 17:05:54 UTC 2022 x86_64 GNU/Linux)
Python: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]
KiCad: 6.0.5+dfsg-1~bpo11+1
Kibot: 1.2.0

Modules:
Colorama: 0.4.4
Distutils: 3.9.2
LXML: 4.6.3
PyYAML: 5.3.1
QRCodeGen: Ok
Requests: 2.25.1
XLSXWriter: 1.1.2

Tools:
Ghostscript: 9.53.3
Git: 2.30.2
ImageMagick: 6.9.11.60 (Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org)
Interactive HTML BoM: 2.5.0 (v2.5.0)
KiBoM: 1.8.0 (KiBOM Version: 1.8.0)
KiCad Automation tools: 1.6.15 (1.6.15 - Copyright 2018-2022, INTI/Productize SPRL - License: Apache)
KiCost: 1.1.10 (KiCost v1.1.10)
Pandoc: Not available
PcbDraw: 0.9.0.3 (PcbDraw 0.9.0-3)
RAR: 5.50.0 (RAR 5.50   Copyright (c) 1993-2017 Alexander Roshal   11 Aug 2017)
RSVG tools: 2.50.3

* Pandoc not installed or too old
  Install the `pandoc` package, i.e.: `sudo apt-get install pandoc`
  You should also install the following packages: texlive-latex-base, texlive-latex-recommended
  - Optional to create PDF/ODF/DOCX files for `report`


Color reference: ok, optional for an output, optional for general use, mandatory for an output, mandatory for general use

Did this help? Please consider commenting it on https://github.com/INTI-CMNB/KiBot/discussions/categories/kibot-check
```

**Comment by set-soft** [comment]
Hi @cdwilson !
Is intentional, you have to use the `setsoft/kicad_auto_test:ki6` image. 
This is because Pandoc (and its dependencies) are huge:
```
setsoft/kicad_auto_test   ki6              e425bbc4e1e2   4 days ago      1.63GB
setsoft/kicad_auto        ki6              6ea0c45a78b5   4 days ago      1.22GB
```

This is almost 400 MB for one tool, the `kicad_auto_test` image has other stuff, but Pandoc is the bigest (by far).

**Reply by set-soft** [reply]
BTW: In the next release the following message will be displayed by `kibot-check`:

 ```
 In CI/CD environments: the `kicad_auto_test` docker image contains it.
```

It was displayed by `kibot` and I forgot to add it to `kibot-check`, so thanks for the reminder ;-)

**Reply by cdwilson** [reply]
@set-soft thanks for the confirmation.  For production environments, is the `kicad_auto_test` docker image recommended to be used?  Or is it only recommended for dev/testing purposes? (trying to understand if the `kicad_auto_test` docker image is always just a superset, or is used for a different purpose)

Also, just a heads up, I ran the kibot-check in the `kicad_auto_test` docker image and it appears to have trouble finding python3 and ibom:

```
# kibot-check
KiBot installation checker

Core:
Linux: 5.10.104 (Linux fdf442307933 5.10.104-linuxkit #1 SMP PREEMPT Thu Mar 17 17:05:54 UTC 2022 x86_64 GNU/Linux)
Python: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]
KiCad: 6.0.5+dfsg-1~bpo11+1
Kibot: 1.2.0

Modules:
Colorama: 0.4.4
Distutils: 3.9.2
LXML: 4.6.3
PyYAML: 5.3.1
QRCodeGen: Ok
Requests: 2.25.1
XLSXWriter: 1.1.2

Tools:
Ghostscript: 9.53.3
Git: 2.30.2
ImageMagick: 6.9.11.60 (Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org)
Failed to run python3, error 1
Interactive HTML BoM: *UNKNOWN*
KiBoM: 1.8.0 (KiBOM Version: 1.8.0)
KiCad Automation tools: 1.6.15 (1.6.15 - Copyright 2018-2022, INTI/Productize SPRL - License: Apache)
KiCost: 1.1.10 (KiCost v1.1.10)
Pandoc: 2.9.2.1
PcbDraw: 0.9.0.3 (PcbDraw 0.9.0-3)
RAR: 5.50.0 (RAR 5.50   Copyright (c) 1993-2017 Alexander Roshal   11 Aug 2017)
RSVG tools: 2.50.3

* Interactive HTML BoM not installed or too old
  Visit: https://github.com/INTI-CMNB/InteractiveHtmlBom
  Download it from: https://github.com/INTI-CMNB/InteractiveHtmlBom/releases
  - Mandatory for `ibom`
```

**Reply by set-soft** [reply]
> @set-soft thanks for the confirmation. For production environments, is the `kicad_auto_test` docker image recommended to be used? 

Yes. `kicad_auto_test` is created from `kicad_auto`, the `ki6` tag is the last KiBot release for both.
Note that you could want to use `ki6.0.5_Debian` tag (or similar) for production. The `ki6` will be updated in the next release, `ki6.0.5_Debian` will remain the same, no updates, so no new features, but no new bugs ;-)

> Or is it only recommended for dev/testing purposes? (trying to understand if the `kicad_auto_test` docker image is always just a superset, or is used for a different purpose)

Is a superset, is used to test KiBot itself, so it contains more tools.

> Also, just a heads up, I ran the kibot-check in the `kicad_auto_test` docker image and it appears to have trouble finding python3 and ibom:

iBoM is a little bit complex, not sure what's the difference, but try defining `export INTERACTIVE_HTML_BOM_NO_DISPLAY=True` 

```shell
export INTERACTIVE_HTML_BOM_NO_DISPLAY=True 
```

This fixes the problem, I tried the current version running on `kicad_auto_test:ki6` and doesn't need it.

**Comment by cdwilson** [comment]
> iBoM is a little bit complex, not sure what's the difference, but try defining `export INTERACTIVE_HTML_BOM_NO_DISPLAY=True`

Thanks, this fixed the issue for me 👍

> I tried the current version running on kicad_auto_test:ki6 and doesn't need it.

Were you running this with X11 forwarding enabled?  I think that may be what was causing the error on my end.  I got the iBoM error above running `kibot-check` on docker for Mac (on Apple Silicon), but I didn't have any X11 forwarding options enabled for the container.

For anybody else that runs across this in the future, here's the `docker run` command I ended up using that fixed the issue:

```
docker run \
    --rm \
    -it \
    --name="KiBot" \
    --env NO_AT_BRIDGE=1 \
    --env INTERACTIVE_HTML_BOM_NO_DISPLAY=True \
    --volume="$(pwd):/home/root/workdir" \
    --workdir="/home/root/workdir" \
    setsoft/kicad_auto_test:ki6 \
    /bin/bash
```

The alternative is to set up X11 forwarding.  (Note: I don't normally do this because KiCad doesn't fully run in GUI mode via X11 forwarding since `inotify` does not work in Intel-based containers on Apple silicon machines.)

```
IP=$(ifconfig en0 | grep inet | awk '$1=="inet" {print $2}')
xhost +$IP
docker run \
    --rm \
    -it \
    --env DISPLAY=$IP:0 \
    --env XAUTHORITY=/.Xauthority \
    --env LIBGL_ALWAYS_INDIRECT=1 \
    --volume="/tmp/.X11-unix:/tmp/.X11-unix" \
    --volume="$HOME/.Xauthority:/.Xauthority" \
    --name="KiBot" \
    --env NO_AT_BRIDGE=1 \
    --volume="$(pwd):/home/root/workdir" \
    --workdir="/home/root/workdir" \
    setsoft/kicad_auto_test:ki6 \
    /bin/bash
```

@set-soft so far, KiBot seems to be working without any X11 forwarding, but I have not exercised all the configuration options.  Do you know if any parts of KiBot require X11 forwarding to be enabled?

**Reply by set-soft** [reply]
> > I tried the current version running on kicad_auto_test:ki6 and doesn't need it.
> 
> Were you running this with X11 forwarding enabled? I think that may be what was causing the error on my end. I got the iBoM error above running `kibot-check` on docker for Mac (on Apple Silicon), but I didn't have any X11 forwarding options enabled for the container.

Yes, you are right, to test the current code I used an script that mounts my home, and it exported X11 stuff.
I added an explicit `INTERACTIVE_HTML_BOM_NO_DISPLAY` definition to the checker. Thanks for pointing this.

> @set-soft so far, KiBot seems to be working without any X11 forwarding, but I have not exercised all the configuration options. Do you know if any parts of KiBot require X11 forwarding to be enabled?

It doesn't need X11 forwarding, all KiCad GUI stuff is executed using a virtual X11 server, so you don't need another X11 server.
